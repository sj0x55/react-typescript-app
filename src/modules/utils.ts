export const compose = <T extends unknown[], R>(fn1: (...args: T) => R, ...fns: Array<(a: R) => R>) => {
  const piped = fns.reduce(
    (prevFn, nextFn) => (value: R) => nextFn(prevFn(value)),
    (val) => val,
  );
  return (...args: T) => piped(fn1(...args));
};
