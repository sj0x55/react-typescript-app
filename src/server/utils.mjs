import terminus from '@godaddy/terminus';
import http from 'http';
import kill from 'kill-port';
import axios from 'axios';
import cheerio from 'cheerio';
import userAgents from './user-agents.mjs';

export async function killPort(port) {
  if (process.env.NODE_ENV === 'development') {
    try {
      await kill(port); // After re-run the server during the watch mode the port may be busy.
    } catch (err) {
      console.error(err);
    }
  }
}

export function getCorsConfig(ipAddress) {
  return {
    origin: (origin, callback) => {
      if (new RegExp(`(http|http)://(localhost|127.0.0.1|${ipAddress})`).test(origin)) {
        callback(null, [origin]);
      }
    },
    methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
  };
}

export async function listenServer(app, port, callback) {
  const server = http.createServer(app);
  const onSignal = async () => {
    console.log('Server is starting cleanup');
  };
  const onShutdown = async () => {
    console.log('Cleanup finished, server is shutting down');
  };

  terminus.createTerminus(server, { signal: 'SIGINT', onSignal, onShutdown, logger: console.error });
  server.listen(port, '0.0.0.0', callback);
}

export async function fetchContent(url) {
  try {
    const userAgent = userAgents[Math.floor(Math.random() * userAgents.length)];
    const result = await axios.get(url, {
      params: {
        t: new Date().getTime(),
      },
      headers: {
        Pragma: 'no-cache',
        Expires: '0',
        'Cache-Control': 'no-cache',
        'User-Agent': userAgent,
      },
    });

    return await cheerio.load(result.data);
  } catch (err) {
    console.error(err);
    return await cheerio.load('');
  }
}

export async function fetchAllDataFromWishList(baseUrl, pathUrl) {
  const parsePrice = (price) => (price && price.trim().replace(/[^0-9.]+/, '')) || 0;
  const prepareProduct = (product) => (condition, price) => ({
    ...product,
    condition,
    price: parseFloat(price),
  });
  const data = [];
  const content = await fetchContent(`${baseUrl}${pathUrl}`);

  content('[id^="itemInfo_"]').each((i, item) => {
    const dataInfoElement = content(item).find('[data-item-prime-info]');
    const { id, asin } = JSON.parse(dataInfoElement.attr('data-item-prime-info') || '{}');

    if (asin) {
      const title = content(item).find(`#itemName_${id}`).text().trim();
      const href = content(item).find(`#itemName_${id}`).attr('href').trim().split('?')[0];
      const priceForNew = parsePrice(content(item).find(`#itemPrice_${id} > span:first-child`).text());
      const priceForUsed = parsePrice(content(item).find('.itemUsedAndNew > .itemUsedAndNewPrice').text());
      const capacity = content(item)
        .find('span:not(:has(*))')
        .map((j, span) => {
          const spanText = content(span).text().replace(/\s+/g, ' ').trim();
          const matchedCapacity = spanText.match(/^Capacity\s?:\s?(\d+)\s?TB/i);

          return (matchedCapacity && parseInt(matchedCapacity[1])) || null;
        })
        .toArray()
        .find(Number);

      if (capacity) {
        const cProduct = prepareProduct({
          url: `${baseUrl}${href}`,
          title,
          capacity,
        });

        if (priceForNew) {
          data.push(cProduct('new', priceForNew));
        }

        if (priceForUsed) {
          data.push(cProduct('used', priceForUsed));
        }
      }
    }
  });

  return data;
}

export const arrayToObject = (propKeyNames, items) => {
  const obj = {};
  const genKey = (item) =>
    [propKeyNames]
      .flat()
      .map((key) => (item[key] || '').replace(/[^a-z0-9]+/gi, '-'))
      .filter((val) => val)
      .join('-');

  items.forEach((item, index) => {
    obj[genKey(item)] = { ...item, index };
  });

  return obj;
};
