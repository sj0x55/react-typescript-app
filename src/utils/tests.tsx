import { FC, ReactElement, ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import testRenderer, { TestRendererOptions } from 'react-test-renderer';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { mainTheme } from 'styles/theme';
import appSliceReducer from 'app/slice';

type RenderOptions = {
  store?: EnhancedStore<{ app: AppState }>;
  [key: string]: unknown;
};

const getWrapper = (
  store: EnhancedStore<{ app: AppState }> = configureStore({
    reducer: {
      app: appSliceReducer,
    },
  }),
): FC => {
  const wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return wrapper;
};

const customRender = (ui: ReactElement, opts?: RenderOptions) => {
  const renderResult = render(ui, { wrapper: getWrapper((opts || {}).store), ...opts });
  const firstChild = renderResult.container.firstChild;

  return { ...renderResult, firstChild };
};

const customCreate = (ui: ReactElement, opts?: TestRendererOptions) => {
  const Wrapper = getWrapper();

  return testRenderer.create(<Wrapper>{ui}</Wrapper>, opts);
};

const mockComponent = function () {
  return ({ children, ...props }: { children: ReactNode }) => {
    return <span {...props}>{children}</span>;
  };
};

export * from '@testing-library/react';
export { customCreate as create };
export { customRender as render };
export { mockComponent };
