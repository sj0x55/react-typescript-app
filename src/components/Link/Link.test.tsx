import { render, create, screen } from 'utils/tests';
import { Link } from '.';

const getLinkElement = () => {
  return <Link href="#">Link</Link>;
};

describe('Link component', () => {
  it('should render a link', () => {
    render(getLinkElement());
    expect(screen.getByRole('link')).toBeInTheDocument();
  });

  it('should match a snapshot', () => {
    const tree = create(getLinkElement());

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
