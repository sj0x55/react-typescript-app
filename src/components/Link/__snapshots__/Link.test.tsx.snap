// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should match a snapshot 1`] = `
.c0 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #66748f;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  className="c0"
  href="#"
>
  Link
</a>
`;
